src/app/webbrowser/Browser.qml

```diff
@@ -589,7 +590,8 @@ BrowserView {
         onCloseTabRequested: internal.closeCurrentTab()
         onToggleBookmark: {
             if (isCurrentUrlBookmarked()) BookmarksModel.remove(tab.url)
-            else internal.addBookmark(tab.url, tab.title, tab.icon)
+            // QtWebEngine icons are provided as e.g. image://favicon/https://duckduckgo.com/favicon.ico
+            else internal.addBookmark(tab.url, tab.title, (tab.icon.toString().substring(0,16) == "image://favicon/") ? tab.icon.toString().substring(16) : tab.icon)
```

Comment:
I feel like something more elegant could be found here... Is there any reason that `tab.icon` couldn't be changed to be appropriate for this use, not requiring all this substring nonsense?

Reply:
The crucial point here is, that QtWebEngine does cache the icons, and while the pages / tabs are open, it is no problem to display give the URL provided by QtWebEngine like `image://favicon/https://duckduckgo.com/favicon.ico` to the QML favicon. (so no extra caching needs to be done by morph, we can directly use the locally saved image)
That is why tab.icon should keep that URL in my opinion.
Only for creating the bookmark (and the history) we need the right part of the URL, so that it will be cached and visible in the bookmarks view without accessing the page.
Doing it with the substring was the most transparent way to me here, of course it is possible to use sth. like "if the scheme or the url is *image* and the host is *favicon*, use its path"

Reply:
`( (UrlUtils.extractScheme(tab.icon) === "image") && (UrlUtils.extractHost(tab.icon) === "host") ) ? UrlUtils.extractPath(tab.icon) : tab.icon` maybe, but UrlUtils.extractPath has not yet been defined in UrlUtils ... At that point I thought it does not look much better so I ended up with the substring, but I'm fine with a regex / sth else too.

Reply:
Completing extractPath and adding `schemeIs(url, expectedScheme)` and `hostIs(url, favicon)` is one way this could be completed which would be more readable. Or, well, forget extractPath and just do schemeIs and hostIs for readability. Would be an improvement.